### PR TITLE
[Admin][Core] User switching

### LINKED ISSUE #3925: Cannot impersonate a Customer
Probably a low priority one, but an interesting issue to consider for future. Since impersonation uses Symfony's standard User framework I'm not sure how easy this would be.

In our model we have a roughly a 50/50 split of Customers who are Users and those who are not.

The workaround for us will be that if we need to impersonate a Customer for and reason then we have to convert them into a User - not the end of the world..

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | closes #3925 |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        if (!$this->authorizationChecker->isGranted(AdminUserInterface::DEFAULT_ADMIN_ROLE)) {
+            throw new HttpException(Response::HTTP_UNAUTHORIZED);
+        }
+
+        $user = $this->userProvider->loadUserByUsername($username);
+        if (null === $user) {
+            throw new HttpException(Response::HTTP_NOT_FOUND);
+        }
+
+        $this->logginer->impersonate($user);
```
Comment (Arminek) [inline src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php]: $this->impersonator?

File: src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+     * @param Request $request
+     * @param string $username
+     *
+     * @return Response
+     */
+    public function impersonateAction(Request $request, $username)
+    {
+        $redirectRoute = $request->attributes->get('redirect');
+
+        if (!$this->authorizationChecker->isGranted(AdminUserInterface::DEFAULT_ADMIN_ROLE)) {
```
Comment (Arminek) [inline src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php]: This should be configurable

File: src/Sylius/Bundle/CoreBundle/Security/UserImpersonator.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    private $sessionTokenParameter;
+
+    /**
+     * @param RequestStack $requestStack
+     * @param string $targetFirewallContextName
+     */
+    public function __construct(RequestStack $requestStack, $targetFirewallContextName)
+    {
+        $this->requestStack = $requestStack;
```
Comment (Arminek) [inline src/Sylius/Bundle/CoreBundle/Security/UserImpersonator.php]: Can you just use `'session'`?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing.yml [hunk 4]
```diff
@@ -92,3 +92,11 @@ sylius_admin_tax_rate:
 
 sylius_admin_zone:
     resource: "@SyliusAdminBundle/Resources/config/routing/zone.yml"
+
+sylius_admin_switch_user_redirect:
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing.yml]: Why do we need such weird route? :D

File: features/admin/impersonating_users.feature [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is an administrator "teddy@roosevelt.com"
+        And there is a customer "Tanith Low" identified by an email "brainsucker@london.uk" and a password "swordlover1918"
+        And I am logged in as "teddy@roosevelt.com" administrator
+
+    @ui
+    Scenario: Impersonating a customer
+        When I view details of the customer "brainsucker@london.uk"
```
Comment (michalmarcinkowski) [inline features/admin/impersonating_users.feature]: `brainsucker` - really?

File: features/admin/impersonating_users.feature [hunk 6]
```diff
@@ -0,0 +1,44 @@
+@user_impersonation
+Feature: Impersonating shop users
+    In order to provide topnotch customer support
+    As an Administrator
+    I want to impersonate a store user
```
Comment (michalmarcinkowski) [inline features/admin/impersonating_users.feature]: **shop** user

File: features/admin/impersonating_users.feature [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    @ui
+    Scenario: Impersonating a customer
+        When I view details of the customer "brainsucker@london.uk"
+        And I try to impersonate them
+        Then I should be logged in as "Tanith Low"
+
+    @ui
+    Scenario: Inability to impersonate a customer with no account
+        Given there is a customer "Harold Zombie" identified by an email "harold@thrasher.ie" and a password "yesmaster"
```
Comment (michalmarcinkowski) [inline features/admin/impersonating_users.feature]: There is a customer with password? We should have a step where we create just customer (not ShopUser) without password.

File: features/admin/impersonating_users.feature [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        And this customer does not have an account
+        When I view their details
+        Then I should be unable to impersonate them
+
+    @ui
+    Scenario: Keeping the administrator access while impersonating a user
+        When I view details of the customer "brainsucker@london.uk"
+        And I try to impersonate them
```
Comment (michalmarcinkowski) [inline features/admin/impersonating_users.feature]: And I ~~try to~~ impersonate them

File: src/Sylius/Behat/Context/Setup/CustomerContext.php [hunk 9]
```diff
@@ -154,6 +154,22 @@ public function thereIsAdministratorIdentifiedByEmailAndPassword($name, $email,
     }
 
     /**
+     * @Given /^(this customer) does not have an account$/
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/CustomerContext.php]: This step should be removed (see above).

File: src/Sylius/Behat/Context/Ui/Admin/SwitchingUsersContext.php [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+use Sylius\Behat\Page\Admin\Customer\ShowPageInterface;
+use Sylius\Behat\Page\Admin\DashboardPageInterface;
+use Sylius\Behat\Page\Shop\HomePageInterface;
+use Sylius\Component\Customer\Model\CustomerInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+final class SwitchingUsersContext implements Context
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/SwitchingUsersContext.php]: We should standardizie vocabulary and always use impersonate customer

File: src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class SwitchUserController
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AdminBundle/Controller/SwitchUserController.php]: `ImpersonateUserController`

File: features/admin/impersonating_users.feature [hunk 12]
```diff
@@ -0,0 +1,44 @@
+@user_impersonation
+Feature: Impersonating shop users
+    In order to provide topnotch customer support
```
Comment (NeverResponse) [inline features/admin/impersonating_users.feature]: top notch, top-notch and topnotch; all of those are valid

File: src/Sylius/Bundle/AdminBundle/Controller/ImpersonateUserController.php [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * @param Request $request
+     * @param string $username
+     */
+    private function addFlash(Request $request, $username)
+    {
+        /** @var Session $session */
+        $session = $request->getSession();
+        $session->getFlashBag()->add('success', $this->translator->trans('sylius.impersonation.success', ['%name%' => $username], 'flashes'));
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AdminBundle/Controller/ImpersonateUserController.php]: I think we should translate the string in the template. You should pass array `['message' => 'sylius.user.impersonate.success', 'parameters' => $username]`.

Btw. discuss translation key of this flash with @GSadee.

**Comment by pjedrzejewski** [comment]
Thank you Jasiek, nice work!

**Pull request merged**